docs/proposals/20220209-machinepool-machines.md

```diff
@@ -0,0 +1,196 @@
+
+As another example, currently each MachinePool infrastructure provider must implement their own node drain functionality. When MachinePool Machines become available in the CAPI API, all providers can use its existing node drain implementation.
+
+### Goals
+
+- Enable CAPI MachinePool to create Machines to represent infrastructure provider MachinePool Machines.
+- Extend CAPI MachinePool to include deployment strategies mirroring those of MachineDeployments
+- Ensure that Machines in MachinePools participate in the same behaviors as MachineDeployment Machines (node drain, machine health checks, retired node deletion)
+
+### Non-Goals/Future Work
+
+- Require infrastructure providers to implement MachinePool Machines.
+- Implement Cluster Autoscaler MachinePool integration (a separate task, although we have a working POC)
+
+## Proposal
+
+To create MachinePool Machines, a MachinePool in CAPI needs information about the instances or replicas associated with the provider's implementation of the MachinePool. This information is attached to the MachinePool as a new optional `infrastructureRefList` field, which is populated by the infrastructure provider.
+
+If the `infrastructureRefList` field is populated, the MachinePool will create MachinePool Machines. A MachinePool Machine is implemented as a Cluster API Machine that is owned by a MachinePool. If the field is empty, no MachinePool Machines will be created, and the overall behavior is the same as before this proposal was implemented.
+
+It is the responsibility of each provider to populate the `infrastructureRefList` field and to create provider-specific MachinePool Machine resources behind each Machine. For example, the Docker provider will create a DockerMachinePoolMachine to represent the container instance behind the Machine in the infrastructure provider's MachinePool.
+
+![MachinePool Controller Infrastructure Reference Changed Sequence](images/machinepool-machines/infraref-changed-sequence.png)
+
+### User Stories
+
+--------------------------------------------
+| ID | Story                               |
+|----|-------------------------------------|
+| U1 | Cluster Autoscaler                  |
+| U2 | MachinePool Machine Remediation     |
+| U3 | MachinePool Machine Rolling Updates |
+--------------------------------------------
+
+#### Story U1
+
+A cluster admin is tasked with using Cluster Autoscaler to scale a MachinePool as required by machine load. The MachinePool should provide a mechanism for Cluster Autoscaler to increase replica count, and provide a mechanism for Cluster Autoscaler to reduce replica count by selecting an optimal machine for deletion.
```

Comment:
>  provide a mechanism for Cluster Autoscaler to reduce replica count by selecting an optimal machine for deletion

is the behavior when a Machine is deleted to reduce the replica count? or to replace that machine with a new one to maintain the desired replica count

Reply:
I'm not sure it's feasible to have multiple source of truths for the replica count (I assume the replica field on the infra MachinePool and potentially the count of current Machines)

Reply:
CA doesn't delete a MachinePool Machine directly, it instead annotates them for deletion and then decrements the replica count on the scalable resource (MachinePool). MachinePool sees the replica count changed, sorts its Machines by that "delete machine" annotation, then deletes the first one.

(Does that clarify? This should be added to the doc, although the mechanism is probably only going to be used by cluster-autoscaler.)

Reply:
MachinePool sees the replica count changed, sorts its Machines by that "delete machine" annotation, then deletes the first one(s).
Will it be done by the machinepool reconciler? What is the purpose if machinepoolmachine reconciler then? When the replica count is decreased, only the machinepool reconciler will be called right? I think the doc should explain what are the new roles and responsibilities of each component, i.e. capi amchinepool controller, infra machinepool reconciler and infra machinepoolmachine reconciler.